updates_and_signals/atomic_message_handlers/workflow.py

```diff
@@ -0,0 +1,198 @@
+import asyncio
+import logging
+import uuid
+from dataclasses import dataclass
+from datetime import timedelta
+from typing import Dict, List, Optional
+
+from temporalio import activity, common, workflow
+from temporalio.client import Client, WorkflowHandle
+from temporalio.common import RetryPolicy
+from temporalio.worker import Worker
+
+from updates_and_signals.atomic_message_handlers.activities import (
+    AllocateNodesToJobInput,
+    DeallocateNodesForJobInput,
+    FindBadNodesInput,
+    allocate_nodes_to_job,
+    deallocate_nodes_for_job,
+    find_bad_nodes,
+)
+
+
+# In workflows that continue-as-new, it's convenient to store all your state in one serializable structure
+# to make it easier to pass between runs
+@dataclass(kw_only=True)
+class ClusterManagerState:
+    cluster_started: bool = False
+    cluster_shutdown: bool = False
+    nodes: Optional[Dict[str, Optional[str]]] = None
+    max_assigned_nodes: int = 0
+    num_assigned_nodes: int = 0
+
+
+@dataclass(kw_only=True)
+class ClusterManagerInput:
+    state: Optional[ClusterManagerState] = None
+    test_continue_as_new: bool = False
+
+
+@dataclass
+class ClusterManagerResult:
+    max_assigned_nodes: int
+    num_assigned_nodes: int
+
+
+# ClusterManagerWorkflow keeps track of the allocations of a cluster of nodes.
+# Via signals, the cluster can be started and shutdown.
+# Via updates, clients can also assign jobs to nodes and delete jobs.
+# These updates must run atomically.
+@workflow.defn
+class ClusterManagerWorkflow:
+    def __init__(self) -> None:
+        self.state = ClusterManagerState()
+        # Protects workflow state from interleaved access
+        self.nodes_lock = asyncio.Lock()
+
+    @workflow.signal
+    async def start_cluster(self):
+        self.state.cluster_started = True
+        self.state.nodes = {str(k): None for k in range(25)}
+        workflow.logger.info("Cluster started")
+
+    @workflow.signal
+    async def shutdown_cluster(self):
+        await workflow.wait_condition(lambda: self.state.cluster_started)
+        self.state.cluster_shutdown = True
+        workflow.logger.info("Cluster shut down")
+
+    @workflow.update
+    async def allocate_n_nodes_to_job(
+        self,
+        task_name: str,
+        num_nodes: int,
+    ) -> List[str]:
+        await workflow.wait_condition(lambda: self.state.cluster_started)
+        if self.state.cluster_shutdown:
+            raise RuntimeError(
```

Comment:
This (and the `ValueError` below) are task failures. You may want to use `ApplicationError`.